### PR TITLE
K8s: API Proxy

### PR DESCRIPTION
**Pull request by toddtreece**
work in progress...

TODO:
* [x] add proxy to http server
* [x] add auth headers to proxied request
* [x] make sure auth headers are parsed properly and added to k8s context

test URL: http://localhost:3000/k8s/apis/dashboard.core.grafana.com/v0-0alpha1/namespaces/default/dashboards?limit=500

### CODE REVIEW

File: pkg/api/k8s_proxy.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+var k8sProxyTransport = &http.Transport{
+	Proxy: http.ProxyFromEnvironment,
+	DialContext: (&net.Dialer{
+		Timeout:   30 * time.Second,
+		KeepAlive: 30 * time.Second,
+	}).DialContext,
+	TLSHandshakeTimeout: 10 * time.Second,
+	TLSClientConfig: &tls.Config{
+		InsecureSkipVerify: true,
```
Comment (DanCech) [inline pkg/api/k8s_proxy.go:28]: don't we need to set `Certificates` or `GetClientCertificate` here to provide the client cert we'll send to k8s?

File: pkg/api/k8s_proxy.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/grafana/grafana/pkg/infra/appcontext"
+	"github.com/grafana/grafana/pkg/infra/log"
+	contextmodel "github.com/grafana/grafana/pkg/services/contexthandler/model"
+	"github.com/grafana/grafana/pkg/util"
+	"github.com/grafana/grafana/pkg/util/proxyutil"
+	"github.com/grafana/grafana/pkg/web"
+)
+
+var k8sProxyTransport = &http.Transport{
+	Proxy: http.ProxyFromEnvironment,
```
[the remark this answers is not in the document]
Comment (toddtreece) [reply pkg/api/k8s_proxy.go:21]: good catch!

File: pkg/api/k8s_proxy.go (hunk 1 above)
Comment (toddtreece) [reply pkg/api/k8s_proxy.go:28]: 🤔 yeah, i think some other config settings on that branch might be hiding this error. i'll look into it

File: pkg/api/k8s_proxy.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+			DialContext: (&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}).DialContext,
+			TLSHandshakeTimeout: 10 * time.Second,
+			TLSClientConfig: &tls.Config{
+				RootCAs:      rootCerts,
+				ClientCAs:    rootCerts,
+				Certificates: []tls.Certificate{cert},
+				ClientAuth:   tls.RequireAndVerifyClientCert,
```
Comment (DanCech) [inline pkg/api/k8s_proxy.go:80]: I don't think we need this or ClientCAs, as those are used server-side and not relevant to the client afaik

File: pkg/services/k8s/apiserver/service.go [hunk 4]
```diff
@@ -130,7 +130,7 @@ func (s *service) writeKubeConfiguration(restConfig *rest.Config) error {
 	clusters := make(map[string]*clientcmdapi.Cluster)
 	clusters["default-cluster"] = &clientcmdapi.Cluster{
 		Server:                restConfig.Host,
-		InsecureSkipTLSVerify: restConfig.Insecure,
+		InsecureSkipTLSVerify: true,
```
Comment (DanCech) [inline pkg/services/k8s/apiserver/service.go:133]: seems like we could disable this if we specify CertificateAuthority or CertificateAuthorityData here using the root ca cert

File: pkg/api/k8s_proxy.go (hunk 3 above)
Comment (toddtreece) [reply pkg/api/k8s_proxy.go:80]: @DanCech i moved this back to draft. i'm trying to debug why [PeerCertificates](https://github.com/kubernetes/kubernetes/blob/bd2404355109fd9f7ca6b14188b4c71c597a9f6f/staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go#L134) is empty here, so i pushed up a WIP commit so @charandas to take a look

File: pkg/services/k8s/apiserver/service.go (hunk 4 above)
Comment (toddtreece) [reply pkg/services/k8s/apiserver/service.go:135]: for some reason i can't get the CA data to work, but that seems like an issue unrelated to this PR. @charandas any ideas?

**Comment by DanCech** [comment]
On both this and the grafana-http-route branch I see this in the server logs:
```
INFO [04-13|06:03:14] No authentication-kubeconfig provided in order to lookup client-ca-file in configmap/extension-apiserver-authentication in kube-system, so client certificate authentication won't work. logger=k8s.apiserver
INFO [04-13|06:03:14] No authentication-kubeconfig provided in order to lookup requestheader-client-ca-file in configmap/extension-apiserver-authentication in kube-system, so request-header client certificate authentication won't work. logger=k8s.apiserver
INFO [04-13|06:03:14] No authorization-kubeconfig provided, so SubjectAccessReview of authorization tokens won't work. logger=k8s.apiserver
INFO [04-13|06:03:14] Neither kubeconfig is provided nor service-account is mounted, so APIPriorityAndFairness will be disabled logger=k8s.apiserver
create storage for GR: customresourcedefinitions.apiextensions.k8s.ioINFO [04-13|06:03:14] Skipping API apiextensions.k8s.io/v1beta1 because it has no resources. logger=k8s.apiserver
create storage for GR: apiservices.apiregistration.k8s.ioINFO [04-13|06:03:14] Skipping API apiregistration.k8s.io/v1beta1 because it has no resources. logger=k8s.apiserver
```
might explain why client cert auth doesn't work...

**Comment by toddtreece** [comment]
> might explain why client cert auth doesn't work...
@DanCech the proxy endpoint should be using client cert auth, which seems to work. how are you testing?

also, did you have an opinion on this approach vs the one in #66417?

**Comment by ryantxu** [comment]
Lets close this one... now that https://github.com/grafana/grafana/pull/66417 is merged

**Pull request closed without merging**